code/socialDistribution/models.py

```diff
@@ -18,6 +18,59 @@ class Author(models.Model):
     username = models.CharField(max_length=50, default='', unique=True)
     displayName = models.CharField(max_length=50)
     githubUrl = models.CharField(max_length=50, null=True)
+    followee = models.ManyToManyField('Author', related_name='followees', blank=True)
+    friend = models.ManyToManyField('Author', related_name='friends', blank=True)
+
+    LOCAL = 'LOCAL'
+    REMOTE = 'REMOTE'
+    HOSTS = [
+        ('LOCAL', 'Local'),
+        ('REMOTE', 'Remote')
+    ]
+
+    host_type = models.CharField(
+        choices=HOSTS,
+        max_length=6,
+        default=LOCAL
+    )
+
+    def is_following(self, author):
+        return self.followee.filter(pk=author.id).exists()
+    
+    def is_friends_with(self, author):
+        return self.friend.filter(pk=author.id).exists()
+
+    def accept_friend(self, author):
+        if author.is_following(self):
+            author.followee.remove(self)
+            self.friend.add(author)
+            author.friend.add(self)
+            return True
+        return False
+
+    def follow(self, author):
+        if author.is_following(self):
+            return self.accept_friend(author)
+
+        if author.id != self.id and not self.is_friends_with(author):
+            # it doesn't matter if self is already following author
+            # django will not create duplicate entries
+            self.followee.add(author)
+            return True
+        return False
+
+    def unfollow(self, author):
+        if author.id != self.id and self.is_following(author):
+            self.followee.remove(author)
+            return True
+        return False
+        
+    def unfriend(self, author):
+        if author.id != self.id and self.is_friends_with(author):
+            self.friend.remove(author)
+            author.friend.remove(self)
+            return True
+        return False
```

Comment:
I'm not sure about the data representation of friends/followers. I believe they should be the same thing, it's just that the user stories are intentionally vague, lol. See https://eclass.srv.ualberta.ca/mod/forum/discuss.php?d=1827396#p4797728 for clarification. 

TLDR: A friend request is equivalent to seeking permission to follow. "Friending" someone is the same as "following" someone. A bidirectional follow is a full friendship. 

For the data representation: I think storing both friends and followers will cause issues with data integrity and data duplication. I think we just need one. Maybe we can have a method on Author that checks for friendship by comparing the follow list of both users?

For the UI: I found the flow of actions a bit confusing with the switching around between follow and befriend. What's the difference between unfriend and unfollow? I think we should pick one terminology (i.e. friending or following) and stick to that consistently.

Sorry for the long comment. Hopefully that all makes sense.

Reply:
Actually, I have an idea. Instead of folloewee/friends, we can have a **follower** field. 

A sends a follow/friend request to B. This is not stored in the Authors db yet, it just goes to B's inbox. Once they accept, A gets added to their "followers". This way, when A makes a post, we send it to Bs inbox since we know B is a follower. If B has A in its followers and B has A in its followers, they are "full friends"

Let me know what you think.